Release Editor Script

sf1919 opened the pull request:
**Description of work.**
During release the separate release notes need to be copied over to the final files. This script does that and moves the separate release notes that have been copied to 'Used' folders to denote they have been copied over. This kick starts the move to manual release notes during release and should give the Release Editor the ability to clean and edit release notes whilst developers continue to add separate release notes as normal.

**To test:**
1. Take a look at the upper level release notes files for v6.4.0 e.g. diffraction.rst. Check that under each heading is a sphinx directive for '''amalgamate'''
2. Check that you have separate release notes in the sub-folders for v.6.4.0 release. None of these folders should contain a folder entitled 'Used'.
3. Navigate to the ../tools/ReleaseNotes folder in your command line
4. Run ```python release_editor.py --release 6.4.0```
5. Check that the upper level release notes file (see 1 above) have either release notes or empty lines instead of the ```amalgamate``` directives
6. Check that the separate release notes (see 2 above) are now in new sub folders called 'Used'
7. Read through the new guidance for release editors and check that the instructions make sense.

Part of #33593 

*This does not require release notes* because **it is a script for developer use only**


---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

sf1919 committed:
Initial version of editor script

sf1919 committed:
Revised release_editor.py script

Part of #33593

sf1919 committed:
Updated ReleaseChecklist for new Release Note process

Part of #33593

ConorMFinn commented on tools/ReleaseNotes/release_editor.py:
```diff
@@ -0,0 +1,126 @@
+
+
+def addReleaseNotesToMain(pathDirectory, listOfDirectories):
+    for file in os.listdir(pathDirectory):
+        if file.endswith(".rst"):
+            with open(mainDirectory + '/' + file) as fileToEdit:
+                for line in fileToEdit:
+                    if line.startswith(directive):
+                        fileName = createFileName(line, pathDirectory, directive)
+                        if fileName in listOfDirectories:
+                            releaseNotes = collateNotes(fileName)
+                        else:
+                            releaseNotes = ""
+                        originalFile = mainDirectory + '/' + file
+                        updateFile(originalFile, releaseNotes, line)
```
This seems a little too nested to be easily readable, could it be refactored?

sf1919 replied:
I wasn't sure about doing the if/else and the listOfDirectories. It could potentially save time but actually the whole thing doesn't take very long anyway so I could probably remove all that which would make it a little less nested.

Not sure if it can be refactored easily otherwise. I had great difficulty accessing the right part of the file to replace.

cailafinn reviewed (changes requested):
Just a quick code comment. Also noticed a couple of functional things:
- The diffraction and workbench output file seems to contain newlines between a few of the notes:
```
- New option to see basic detector data in a tooltip when hovering.

- It is now possible to replace the workspace and save the image of the instrument viewer using its python API.
```
- The notes seem to be collated with two empty lines above and no empty lines below under each heading. 

Neither of these are that significant, and could just be tided up by the editor, but it would be nice to have it be automatic.

sf1919 commented on dev-docs/source/ReleaseChecklist.rst:
```diff
@@ -138,7 +138,18 @@ language, layout, and collecting images.
+*  Switch to manual handling of release notes by running the release_editor.py script using the
+   `release editor helper tool
+   <https://github.com/mantidproject/mantid/blob/main/tools/ReleaseNotes/release_editor.py>`_
+   and open a pull request to put them on ``release``. The script copies all of separate release notes under the correct heading of their upper level file, e.g. framework.rst, and moves
```
I can put in a semi-fix but the empty lines around the headings are partially set by where the amalgamate directive sits within the page. The script replaces that single line with the amalgamate directive in it with the release notes.

sf1919 committed:
Remove unneccessary empty lines from collation of notes

sf1919 committed:
Refactor and comment addReleaseNotesToMain

sf1919 committed:
Update dev-docs/source/ReleaseChecklist.rst

Co-authored-by: Conor Finn <47181718+ConorMFinn@users.noreply.github.com>

ConorMFinn commented on tools/ReleaseNotes/release_editor.py:
```diff
@@ -0,0 +1,128 @@
+
+    args.release = fixReleaseName(args.release)
+    mainDirectory = os.path.normpath(createFileLocation(args.release))
+    directive = ".. amalgamate:: "
```
Since this is a constant, this should probably be in all caps

cailafinn reviewed (changes requested):
Sorry, just one more thing I missed from before

sf1919 committed:
Change constant to all caps and move to top of file

cailafinn reviewed (dismissed):
Looks good to me and works well, nice work 👍

martyngigg commented on dev-docs/source/ReleaseChecklist.rst:
```diff
@@ -138,7 +138,18 @@ language, layout, and collecting images.
+*  Switch to manual handling of release notes by running the release_editor.py script using the
+   `release editor helper tool
+   <https://github.com/mantidproject/mantid/blob/main/tools/ReleaseNotes/release_editor.py>`_
+   and open a pull request to put them on ``release``. The script copies all of the separate release notes under the correct heading of their upper level file, e.g. framework.rst, and moves
```
Just a minor correction here as I assume ``release`` should be the name of the release branch: ``release-next``?

martyngigg commented on dev-docs/source/ReleaseChecklist.rst:
```diff
@@ -148,6 +159,11 @@ Tuesday, 4 days
+After Release
+-------------
+
+* Remove all separate release note files and sub-file structure to leave just the upper level release notes e.g. diffraction.rst, index.rst, framework.rst etc.
```
Would it be possible to do this as one of the final steps in preparing to tag the release instead of after? It leaves the source for the release in a clean state then.

martyngigg commented on tools/ReleaseNotes/release_editor.py:
```diff
@@ -0,0 +1,129 @@
+
+def updateFile(originalFile, replaceText, textToReplace):
+    # opens the upper level release note e.g. diffraction.rst
+    mainRST = open(originalFile, 'r')
```
Could I ask that we use the [with](https://realpython.com/python-with-statement/#the-with-statement-approach) statement for managing files, like is done below?

sf1919 replied:
I've changed all instances like this to usng the with statement.

martyngigg commented on tools/ReleaseNotes/release_editor.py:
```diff
@@ -0,0 +1,129 @@
+DIRECTIVE = ".. amalgamate:: "
+
+
+def getReleaseRoot():
```
I know the syle here roughly follows the `release.py` script but that is a little old now. Could we modernize a bit and use [pathlib](https://realpython.com/python-pathlib/#creating-paths) when dealing with paths? For example,

```python
def getReleaseRoot() -> pathlib.Path:
    program_path = pathlib.Path(sys.argv[0])
    return program_path.resolve('../../docs/source/release')
```

This will mean we stop treating paths as strings and allows Python internally to worry about the differences betwen `\` and `/` on Windows. This function now returns a `Path` and not a `str` so I've snuck in a `-> pathlib.Path` on the end of the function to indicate this. This is, [type annotation](https://realpython.com/python-type-checking/#annotations) a good habit for more Python too as it makes it easier to read what types. are expected

sf1919 replied:
Moving to using pathlib is a good idea. However it has required quite a bit of re-writing to make the whole script work. That being said I think it looks much better now.

For this one I don't think .resolve() takes any arguments so I had to include an extra step to get us where we needed.

martyngigg commented on tools/ReleaseNotes/release_editor.py:
```diff
@@ -0,0 +1,129 @@
+
+
+def createFileLocation(releaseNo):
+    release_root = getReleaseRoot()
```
If we make the `pathllib` modification then this can be simply:

```python
def createFileLocation(releaseNo) -> pathlib.Path:
    return getReleaseRoot() / releaseNo
```

`pathlib` means we can use `/` but Python handles using a `\` on Windows paths.

martyngigg commented on tools/ReleaseNotes/release_editor.py:
```diff
@@ -0,0 +1,129 @@
+
+def addReleaseNotesToMain(pathDirectory):
+    # iterates through files in a directory
+    for file in os.listdir(pathDirectory):
```
May I suggest using the [glob](https://docs.python.org/3.8/library/glob.html) module here? It would mean you only get a list of `.rst` files back and don't have to check that yourself: `for filename in glob.glob(pathDirectory.resolve('*.rst')):`

martyngigg reviewed (changes requested):
Just a few comments on Python things that were possibly not known about.

I've been a little pedantic here as I realise the old script did none of this but it would be great to use some of the more modern Python features.

sf1919 committed:
Revision after gatekeeper comments

- moved to using pathlib
- ensured opening and writing files using with statement
- updated Release Checklist

martyngigg commented on tools/ReleaseNotes/release_editor.py:
```diff
@@ -6,15 +6,17 @@
-def getReleaseRoot():
-    script_dir = os.path.split(sys.argv[0])[0]
-    return os.path.abspath(os.path.join(script_dir, '../../docs/source/release/'))
+def getReleaseRoot() -> pathlib.Path:
```
I just realised when testing this out that this has a built-in assumption that the current directory is the same as this script. I know the old script did this but it's a good habit to get into to try and make it so you can be in any directory, i.e. this failed for me

```sh
$ python tools/ReleaseNotes/release_editor.py --release 6.4.0
```

with the following error:

```sh
Traceback (most recent call last):
  File "tools/ReleaseNotes/release_editor.py", line 123, in <module>
    directoriesUsed = checkContainsReleaseNote(mainDirectory)
  File "tools/ReleaseNotes/release_editor.py", line 87, in checkContainsReleaseNote
    getReleaseNoteDirectories(path, releaseNoteList)
  File "tools/ReleaseNotes/release_editor.py", line 73, in getReleaseNoteDirectories
    if len([f for f in os.listdir(path) if f.endswith('.rst')]) > 0:
FileNotFoundError: [Errno 2] No such file or directory: '/Users/dmn58364/Code/git/docs/source/release/v6.4.0'
```

Could we swap out using `sys.argv` for `__file__. The latter variable will always contain a full path to this file wherever it is on the system. We can then be confident of the relative path of the other directories.

martyngigg reviewed (changes requested):
This looks good. Running from the same directory as the script still produces what we expect. I've just got a minor comment about making the current directory slightly more flexible.

sf1919 committed:
Changed file location call

martyngigg reviewed (approved):
I can now run this script without changing to the tools directory. Good stuff! 🎉
